File: app/src/main/java/im/tox/antox/callbacks/AntoxOnConnectionStatusCallback.java
package im.tox.antox.callbacks;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import java.util.ArrayList;

import im.tox.antox.data.AntoxDB;
import im.tox.antox.tox.ToxSingleton;
import im.tox.antox.utils.AntoxFriend;
import im.tox.antox.utils.Constants;
import im.tox.antox.utils.Message;
import im.tox.jtoxcore.ToxException;
import im.tox.jtoxcore.callbacks.OnConnectionStatusCallback;

public class AntoxOnConnectionStatusCallback implements OnConnectionStatusCallback<AntoxFriend> {

    private final static String TAG = "im.tox.antox.TAG";
    private Context ctx;

    ToxSingleton toxSingleton = ToxSingleton.getInstance();

    public AntoxOnConnectionStatusCallback(Context ctx) {
        this.ctx = ctx;
    }

    @Override
    public void execute(AntoxFriend friend, boolean online) {
        AntoxDB db = new AntoxDB(ctx);
        db.updateUserOnline(friend.getId(), online);
        String tmp = online ? "come online" : "gone offline";
        db.addMessage(-1, friend.getId(), friend.getNickname() + " has " + tmp, true, true, true, 5);
        db.close();
        if (online) {
            toxSingleton.sendUnsentMessages(ctx);
        }
        toxSingleton.updateMessages(ctx);
    }
}

File: app/src/main/scala/im/tox/antox/callbacks/AntoxOnConnectionStatusCallback.scala
package im.tox.antox.callbacks

import android.content.{SharedPreferences, Context}
import android.net.ConnectivityManager
import android.preference.PreferenceManager
import im.tox.antox.data.{State, AntoxDB}
import im.tox.antox.tox.{MessageHelper, Reactive, ToxSingleton}
import im.tox.antox.utils.{ConnectionTypeChangeListener, ConnectionManager}
import im.tox.tox4j.core.callbacks.FriendConnectionStatusCallback
import im.tox.tox4j.core.enums.ToxConnection

import scala.collection.JavaConversions._

object AntoxOnConnectionStatusCallback {

  private val TAG = "im.tox.antox.TAG"
}

class AntoxOnConnectionStatusCallback(private var ctx: Context) extends FriendConnectionStatusCallback {

  private val preferences = PreferenceManager.getDefaultSharedPreferences(ctx)
  private var preferencesListener: SharedPreferences.OnSharedPreferenceChangeListener = _

  def setAllStatusNone(): Unit = {
    if (!ToxSingleton.isToxConnected(preferences, ctx)) {
      for (friend <- ToxSingleton.getAntoxFriendList.all()) {
        friendConnectionStatus(friend.getFriendNumber, ToxConnection.NONE)
      }
    }
  }

  ConnectionManager.addConnectionTypeChangeListener(new ConnectionTypeChangeListener {
    override def connectionTypeChange(connectionType: Int): Unit = {
      setAllStatusNone()
    }
  })

  preferencesListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
    override def onSharedPreferenceChanged(prefs: SharedPreferences, key: String): Unit = {
      key match {
        case "wifi_only" =>
          setAllStatusNone()
        case _ =>
      }
    }
  }

  preferences.registerOnSharedPreferenceChangeListener(preferencesListener)


  override def friendConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection): Unit = {
    val online = connectionStatus != ToxConnection.NONE

    val db = new AntoxDB(ctx)
    val friendKey = ToxSingleton.getAntoxFriend(friendNumber).get.getKey
    db.updateContactOnline(friendKey, online)
    ToxSingleton.getAntoxFriend(friendNumber).get.setOnline(online)

    if (online) {
      MessageHelper.sendUnsentMessages(ctx)
      State.transfers.updateSelfAvatar(ctx)
    } else {
      ToxSingleton.typingMap.put(friendKey, false)
      Reactive.typing.onNext(true)
    }
    ToxSingleton.updateFriendsList(ctx)
    ToxSingleton.updateMessages(ctx)

    db.close()
  }
}
